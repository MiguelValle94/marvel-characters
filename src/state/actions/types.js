export const FETCH_CHARACTERS_START = '[Characters] Fetch start';
export const FETCH_CHARACTERS_SUCCESS = '[Characters] Fetch success';
export const FETCH_CHARACTERS_ERROR = '[Characters] Fetch error';
export const FETCH_CHARACTER_DETAILS_START = '[Character Details] Fetch start';
export const FETCH_CHARACTER_DETAILS_SUCCESS =
  '[Character Details] Fetch success';
export const FETCH_CHARACTER_DETAILS_ERROR = '[Character Details] Fetch error';
export const UPDATE_FILTERS = '[Filters] Update';
