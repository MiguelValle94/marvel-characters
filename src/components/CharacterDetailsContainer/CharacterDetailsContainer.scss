@import '../../styles/variables.scss';

.character-details-container {
  &__wrapper {
    display: flex;
    flex-direction: column;
    padding: 0 25px;
  }

  &__back-button {
    align-self: flex-start;
    margin: var(--space-2) 0;
  }
}
