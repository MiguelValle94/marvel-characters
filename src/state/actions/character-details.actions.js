import {
  FETCH_CHARACTER_DETAILS_START,
  FETCH_CHARACTER_DETAILS_SUCCESS,
  FETCH_CHARACTER_DETAILS_ERROR
} from './types';
import {
  MARVEL_API_URL,
  CHARACTERS_ENDPOINT,
  API_KEY
} from '../../constants/config.constants';
import { CUSTOM_ERROR_MESSAGE } from '../../constants/error-messages.constants';

export function fetchCharacterDetailsStart() {
  return {
    type: FETCH_CHARACTER_DETAILS_START
  };
}

export function fetchCharacterDetailsSuccess(payload) {
  return {
    type: FETCH_CHARACTER_DETAILS_SUCCESS,
    payload
  };
}

export function fetchCharacterDetailsError(error) {
  return {
    type: FETCH_CHARACTER_DETAILS_ERROR,
    error
  };
}

export const fetchCharacterDetails = (id) => (dispatch) => {
  dispatch(fetchCharacterDetailsStart());
  return fetch(
    `${MARVEL_API_URL}${CHARACTERS_ENDPOINT}/${id}?apikey=${API_KEY}`
  )
    .then((response) => response.json())
    .then(({ data, code, status }) => {
      if (code >= 400 && code < 500) {
        throw status;
      }
      dispatch(fetchCharacterDetailsSuccess(data));
    })
    .catch((error) =>
      dispatch(
        fetchCharacterDetailsError({
          message:
            error && typeof error === 'string' ? error : CUSTOM_ERROR_MESSAGE
        })
      )
    );
};
