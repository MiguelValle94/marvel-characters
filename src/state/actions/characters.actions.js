import {
  FETCH_CHARACTERS_START,
  FETCH_CHARACTERS_SUCCESS,
  FETCH_CHARACTERS_ERROR
} from './types';
import {
  MARVEL_API_URL,
  CHARACTERS_ENDPOINT,
  API_KEY
} from '../../constants/config.constants';

export function fetchCharactersStart() {
  return {
    type: FETCH_CHARACTERS_START
  };
}

export function fetchCharactersSuccess(payload) {
  return {
    type: FETCH_CHARACTERS_SUCCESS,
    payload
  };
}

export function fetchCharactersError(error) {
  return {
    type: FETCH_CHARACTERS_ERROR,
    error
  };
}

export const fetchCharacters = (orderBy) => (dispatch) => {
  dispatch(fetchCharactersStart());
  return fetch(
    `${MARVEL_API_URL}${CHARACTERS_ENDPOINT}?apikey=${API_KEY}&orderBy=${orderBy}`
  )
    .then((response) => response.json())
    .then(({ data }) => dispatch(fetchCharactersSuccess(data)))
    .catch((error) => dispatch(fetchCharactersError(error)));
};
