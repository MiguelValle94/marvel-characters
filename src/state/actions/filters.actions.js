import { UPDATE_FILTERS } from './types';

export function updateFilters(payload) {
  return {
    type: UPDATE_FILTERS,
    payload
  };
}
