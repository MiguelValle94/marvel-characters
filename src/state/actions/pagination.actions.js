import { SELECT_LIMIT, SELECT_PAGE, UPDATE_PAGINATION } from './types';

export function selectLimit(limit) {
  return {
    type: SELECT_LIMIT,
    payload: limit
  };
}

export function selectPage(page) {
  return {
    type: SELECT_PAGE,
    payload: page
  };
}

export function updatePagination(payload) {
  return {
    type: UPDATE_PAGINATION,
    payload
  };
}
