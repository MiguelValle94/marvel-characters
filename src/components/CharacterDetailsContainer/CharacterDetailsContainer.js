import React, { useEffect } from 'react';
import { connect } from 'react-redux';
import { useParams, Link } from 'react-router-dom';

import { fetchCharacterDetails } from '../../state/actions/character-details.actions';
import CharacterDetails from './CharacterDetails';

import './CharacterDetailsContainer.scss';

function CharacterDetailsContainer({ character, dispatch }) {
  let { id } = useParams();

  useEffect(() => {
    dispatch(fetchCharacterDetails(id));
  }, [dispatch, id]);

  return !character ? (
    <p>Loading...</p>
  ) : (
    <div className='character-details-container__wrapper'>
      <Link to='/' className='character-details-container__back-button'>
        <i className='fa fa-arrow-left'></i>{' '}
        All characters
      </Link>
      <CharacterDetails character={character} />
    </div>
  );
}

function getCharacterFromDetails(state) {
  return state.characterDetails.characterDetails;
}

function getCharacterFromList(state, id) {
  return state.characters.characters.find(
    (character) => character.id.toString() === id
  );
}

const mapStateToProps = (state, ownProps) => ({
  character:
    getCharacterFromDetails(state) ||
    getCharacterFromList(state, ownProps.match.params.id),
  isLoading: state.characterDetails.isLoading
});

export default connect(mapStateToProps)(CharacterDetailsContainer);
