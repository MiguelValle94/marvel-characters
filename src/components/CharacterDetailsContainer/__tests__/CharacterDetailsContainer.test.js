import React from 'react';
import { mount } from 'enzyme';

import CharacterDetailsContainer from '../CharacterDetailsContainer';
import { mockStore } from '../../../state/__mocks__/store';
import { fetchCharacterDetails } from '../../../state/actions/character-details.actions';

jest.mock('../CharacterDetails', () => () => <></>);
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useParams: () => ({ id: 1011334 })
}));
jest.mock('../../../state/actions/character-details.actions', () => ({
  ...jest.requireActual('../../../state/actions/character-details.actions'),
  fetchCharacterDetails: jest.fn()
}));

let store;

const setUpMount = (initialState, match) => {
  store = mockStore(initialState);
  store.dispatch = jest.fn();
  const component = mount(
    <CharacterDetailsContainer store={store} match={match} />
  );
  return component;
};

describe('CharacterDetailsContainer', () => {
  let component;
  const id = 1011334;

  it('should display loading message while no character', () => {
    component = setUpMount(
      {
        charactersReducer: { characters: [] },
        characterDetailsReducer: {
          characterDetails: undefined,
          isLoading: true
        }
      },
      { params: {} }
    );
    expect(component).toMatchSnapshot();
  });

  it('should display character details', () => {
    component = setUpMount(
      {
        charactersReducer: { characters: [] },
        characterDetailsReducer: { characterDetails: [{ id }] }
      },
      { params: { id: id.toString() } }
    );
    expect(component).toMatchSnapshot();
  });

  it('should fetch character details with the id in the url', () => {
    component = setUpMount(
      {
        charactersReducer: { characters: [] },
        characterDetailsReducer: { characterDetails: undefined }
      },
      { params: { id: id.toString() } }
    );
    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith(fetchCharacterDetails(id));
  });

  it('should not fetch character details if already in the store', () => {
    component = setUpMount(
      {
        charactersReducer: { characters: [{ id }] },
        characterDetailsReducer: { characterDetails: undefined }
      },
      { params: { id: id.toString() } }
    );
    expect(store.dispatch).not.toHaveBeenCalled();
  });
});
