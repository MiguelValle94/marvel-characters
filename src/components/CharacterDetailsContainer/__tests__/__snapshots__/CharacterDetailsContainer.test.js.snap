// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterDetailsContainer should display character details 1`] = `
<Connect(CharacterDetailsContainer)
  match={
    Object {
      "params": Object {
        "id": "1011334",
      },
    }
  }
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [MockFunction],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <CharacterDetailsContainer
    character={
      Array [
        Object {
          "id": 1011334,
        },
      ]
    }
    dispatch={[MockFunction]}
    match={
      Object {
        "params": Object {
          "id": "1011334",
        },
      }
    }
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [MockFunction],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  >
    <Component
      character={
        Array [
          Object {
            "id": 1011334,
          },
        ]
      }
    />
  </CharacterDetailsContainer>
</Connect(CharacterDetailsContainer)>
`;

exports[`CharacterDetailsContainer should display loading message while no character 1`] = `
<Connect(CharacterDetailsContainer)
  match={
    Object {
      "params": Object {},
    }
  }
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [MockFunction] {
        "calls": Array [
          Array [
            undefined,
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      },
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <CharacterDetailsContainer
    dispatch={
      [MockFunction] {
        "calls": Array [
          Array [
            undefined,
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
    isLoading={true}
    match={
      Object {
        "params": Object {},
      }
    }
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [MockFunction] {
          "calls": Array [
            Array [
              undefined,
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        },
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  >
    <p>
      Loading...
    </p>
  </CharacterDetailsContainer>
</Connect(CharacterDetailsContainer)>
`;
